nmsg/kafkaio.c

```diff
@@ -0,0 +1,569 @@
+		_kafka_flush(ctx);
+	}
+
+	/* Destroy topic */
+	if (ctx->topic != NULL)
+		rd_kafka_topic_destroy(ctx->topic);
+
+	/* Destroy handle */
+	if (ctx->handle != NULL)
+		rd_kafka_destroy(ctx->handle);
+
+	if (ctx->topic_str != NULL)
+		my_free(ctx->topic_str);
+
+	if (ctx->broker != NULL)
+		my_free(ctx->broker);
+
+	my_free(ctx);
+}
+
+static void
+_kafka_error_cb(rd_kafka_t *rk, int err, const char *reason, void *opaque)
+{
+	rd_kafka_resp_err_t err_kafka = (rd_kafka_resp_err_t) err;
+
+	switch(err_kafka) {
+		case RD_KAFKA_RESP_ERR__UNKNOWN_PARTITION:
+		case RD_KAFKA_RESP_ERR_UNKNOWN_TOPIC_OR_PART:
+		case RD_KAFKA_RESP_ERR_OFFSET_OUT_OF_RANGE:
+		/* At the moment treat any broker's error as fatal */
+		default:
+			_nmsg_dprintf(2, "%s: got Kafka error %d: %s\n", __func__, err, reason);
+			break;
+	}
+}
+
+/* Export. */
+
+void
+kafka_ctx_destroy(kafka_ctx_t *ctx)
+{
+	if (ctx != NULL && *ctx != NULL) {
+		_kafka_ctx_destroy(*ctx);
+		*ctx = NULL;
+	}
+}
+
+nmsg_res
+kafka_read_start(kafka_ctx_t ctx, uint8_t **buf, size_t *len)
+{
+	nmsg_res res = nmsg_res_success;
+	if (buf == NULL || len == NULL || ctx == NULL || !ctx->consumer)
+		return nmsg_res_failure;
+
+	*buf = NULL;
+	*len = 0;
+
+	if (ctx->group_id != KAFKA_GROUP_ID_NONE)
+		ctx->message = rd_kafka_consumer_poll(ctx->handle, ctx->timeout);
+	else {
+		/* Poll for errors, etc. */
+		rd_kafka_poll(ctx->handle, 0);
+		ctx->message = rd_kafka_consume(ctx->topic, ctx->partition, ctx->timeout);
+	}
+
+	if (ctx->message != NULL) {
+		if (ctx->message->err == RD_KAFKA_RESP_ERR_NO_ERROR) {
+			*buf = ctx->message->payload;
+			*len = ctx->message->len;
+		} else {
+			if (ctx->message->err == RD_KAFKA_RESP_ERR__PARTITION_EOF)
+				res = nmsg_res_again;
+			else {
+				_kafka_error_cb(ctx->handle, ctx->message->err,
+						rd_kafka_message_errstr(ctx->message), ctx);
+				res = nmsg_res_failure;
+			}
+
+			/* Return error message to kafka */
+			rd_kafka_message_destroy(ctx->message);
+			ctx->message = NULL;
+		}
+	} else
+		res = (errno == ETIMEDOUT ? nmsg_res_again : nmsg_res_failure);
+
+	return res;
+}
+
+nmsg_res
+kafka_read_close(kafka_ctx_t ctx)
+{
+	if (ctx == NULL || !ctx->consumer)
+		return nmsg_res_failure;
+
+	if (ctx->message != NULL)
+		/* Return message to rdkafka */
+		rd_kafka_message_destroy(ctx->message);
+
+	return nmsg_res_success;
+}
+
+nmsg_res
+kafka_write(kafka_ctx_t ctx, const uint8_t *buf, size_t len)
+{
+	int res;
+
+	if (ctx == NULL || ctx->consumer)
+		return nmsg_res_failure;
+
+	res = rd_kafka_produce(ctx->topic, ctx->partition, RD_KAFKA_MSG_F_FREE,
+			       (void *) buf, len,	/* Payload and length */
+			       NULL, 0,			/* Optional key and its length */
+			       NULL);			/* Opaque data in message->_private. */
+
+	/* Poll to handle delivery reports */
+	rd_kafka_poll(ctx->handle, 0);
+	if (res < 0) {
+		_nmsg_dprintf(1, "%s: failed to produce Kafka message #%d: %s\n",
+			__func__, errno, rd_kafka_err2str(errno));
+		return nmsg_res_failure;
```

Comment:
Note that one of the failure modes of `rd_kafka_produce` is when the output queue is full -- this needs to be handled in a way other than returning an error and shutting down the output.

The best way I can see to handle this is to, when `rd_kafka_produce` fails due to a full output queue, poll for a delivery report and retry.  As with zeromq outputs, the poll should use `NMSG_RBUF_TIMEOUT`, and this loop will need to be canceled by `nmsg_output_stop`.